Realm.Shared/Dynamic/MetaRealmObject.cs

```diff
@@ -0,0 +1,241 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections.Generic;
+using System.Dynamic;
+using System.Linq.Expressions;
+using System.Reflection;
+
+namespace Realms.Dynamic
+{
+    internal class MetaRealmObject : DynamicMetaObject
+    {
+        private readonly Realm _realm;
+        private readonly RealmObject.Metadata _metadata;
+
+        private static readonly FieldInfo RealmObjectRealmField = typeof(RealmObject).GetTypeInfo().GetField("_realm", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+        private static readonly FieldInfo RealmObjectRowHandleField = typeof(RealmObject).GetTypeInfo().GetField("_rowHandle", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+
+        public MetaRealmObject(Expression expression, DynamicRealmObject value)
+            : base(expression, BindingRestrictions.Empty, value)
+        {
+            _realm = value.Realm;
+            _metadata = value.ObjectMetadata;
+        }
+
+        public override DynamicMetaObject BindGetMember(GetMemberBinder binder)
+        {
+            var property = _metadata.Schema.Find(binder.Name);
+            if (property == null)
+            {
+                return base.BindGetMember(binder);
+            }
+
+            var arguments = new List<Expression>
+            {
+                WeakConstant(_metadata.Table),
+                Expression.Field(GetLimitedSelf(), RealmObjectRowHandleField),
+                Expression.Constant(_metadata.ColumnIndices[property.Name])
+            };
+
+            MethodInfo realmObjectOpsMethod = null;
+
+            switch (property.Type)
+            {
+                case Schema.PropertyType.Int:
+                    if (property.IsNullable)
+                        realmObjectOpsMethod = typeof(RealmObjectOps).GetMethod(nameof(RealmObjectOps.GetNullableInt64Value));
```

Comment:
The way `dynamic` works in C# already stashes the result of this computation in a static field. Whenever the compiler encounters a dynamic expression, it creates a `CallSite` object and stores it in a static field. The various `CallSiteBinder` objects such as `GetMemberBinder` and `SetMemberBinder` describe the operation and a series of `DynamicMetaObject`objects produce an expression tree to execute the operation. The `CallSite` compiles this expression tree into a lambda and caches it.

In essence, in
```csharp
for (var i = 1; i <= 1000; i++)
{
    var person = realm.CreateObject("Person");
    person.Name = $"Person {i}";
}
```
`MetaRealmObject.BindSetMember` will be called only once. This is why I think we won't gain much by caching here. Alternatively, we could use something like [`InfoOf.Fody`](https://github.com/Fody/InfoOf), but I didn't feel like pulling in the dependency just for that.